grpc-client-server/README.md

```diff
@@ -0,0 +1,188 @@
+prober_server to accept a frequency argument instead of the old `repetitions` parameter, 
+and continue running probes every N seconds forever. 
+
+You will need to use a [goroutine](https://gobyexample.com/goroutines) to keep running probes in the background.
+
+Now, add a new client operation to get the value of the last probe run.
+To make this work, you will need to pass a value to your prober goroutine that it can use to store the most recent latency.
+You can store this value in your `server` struct.
+You will also need to use a `sync.Mutex` to guarantee that the goroutine that serves your client request sees the last value 
+written by the prober goroutine. A Tour of Go has a good example of [using a sync.Mutex](https://go.dev/tour/concurrency/9).
+
+
+### Part 3: Add Prometheus Metrics
+If you still have time, then let's learn something about how to monitor applications.
+
+In software operations, we want to know what our software is doing and how it is performing.
+One very useful technique is to have your program export metrics. Metrics are basically values that your 
+program makes available (via HTTP). Specialised programs, such as Prometheus, can then fetch metrics regularly
+from all the running instances of your program, store the history of these metrics, and do useful arithmetic on them
+(like computing rates, averages, and maximums). You can use this data to do troubleshooting and to alert if things 
+go wrong.
+
+Read the [Overview of Prometheus](https://prometheus.io/docs/introduction/overview/).
+
+Now add Prometheus metrics to your prober server. Every time you execute a probe, update a `gauge` metric that tracks the latency.
+Add a `label` specifying the endpoint being probed. 
+The [Prometheus Guide to Instrumenting a Go Application](https://prometheus.io/docs/guides/go-application/) has all the information you need to do this.
+
+Once you've run your program, use your client to execute periodic probes against some endpoint. 
+Now use the `curl` program or your browser to view the metrics on the port that you've chosen to serve them on.
+You should see a number of built-in Go metrics, plus your new gauge.
+
+If you use your client to start probing a second endpoint, you should see a second labelled metric appear.
+
+### Part 3: Scrape Prometheus Metrics
+The final step is to set up the Prometheus application to periodically pull metrics from your `prober_server`.
+
+You'll need to set up a simple configuration to scrape your `prober_server`. Prometheus [configuration](https://prometheus.io/docs/prometheus/latest/configuration/configuration/) is quite complex but you can adapt this [example configuration](https://github.com/prometheus/prometheus/blob/main/documentation/examples/prometheus.yml).
+
+Run Prometheus in Docker locally - see [Prometheus Installation](https://prometheus.io/docs/prometheus/latest/installation/).
+Use a shared volume for the configuration file, as suuggested in the guide above.
```

Comment:
The projects thus far have either run entirely in docker, or not in docker at all. I think it's worth introducing here that it's possible (and can be easier?) to run systems like this in docker, rather than install them on your machine.